L74-java-channel-creds.md

```diff
@@ -0,0 +1,478 @@
+   * There may be optional configuration that can be ignored if not understood.
+   */
+  public Set<Feature> incomprehensible(EnumSet<Feature> understoodFeatures) {...}
+
+  /* Other methods can be added to support various Features */
+
+  public enum Feature {}
+
+  public final static class Builder {
+    public TlsChannelCredentials build() {...}
+  }
+}
+```
+
+To support very advanced use cases, Netty will provide a credential that wraps
+a `ProtocolNegotiator`. This allows implementations like ALTS and XDS to use
+internal APIs without forcing their users to use experimental or internal APIs,
+as their users would just interact with `ChannelCredentials`.
+
+```java
+package io.grpc.netty;
+
+final class NettyChannelCredentials extends ChannelCredentials {
+  public NettyChannelCredentials(ProtocolNegotiator.Factory negotiator) {...}
+  public ProtocolNegotiator.Factory getNegotiator() { return negotiator; }
+}
+
+@Internal
+public final class InternalNettyChannelCredentials {
+  private InternalNettyChannelCredentials() {}
+
+  public ChannelCredentials create(InternalProtocolNegotiator.Factory negotiator)
+    {...}
+
+  /**
+   * Converts credentials to a negotiator, in similar fashion as for a new
+   * channel.
+   *
+   * @throws IllegalArgumentException if unable to convert
+   */
+  public InternalProtocolNegotiator.Factory toNegotiator(ChannelCredentials creds)
+    {...}
+}
+```
+
+### GoogleDefault ChannelCredentials Example
+
+The API provided here is not part of the design. It is meant as a
+demonstration.
+
+```java
+public final class GoogleDefaultChannelCredentials {
+  private GoogleDefaultChannelCredentials() {}
+
+  public static ChannelCredentials create() {
+    return new ChoiceChannelCredentials(
+        InternalNettyChannelCredentials.create(
+          new AltsProtocolNegotiatorFactory()),
+        new CompositeChannelCredentials(
+          new TlsChannelCredentials(),
+          MoreCallCredentials.from(GoogleCredentials.getApplicationDefault())));
+  }
+}
+```
+
+### xDS ChannelCredentials Example
+
+The API provided here is not part of the design. It is meant as a
+demonstration.
+
+```java
+public final class XdsChannelCredentials {
+  private XdsChannelCredentials() {}
+
+  /**
+   * Creates credentials to be configured by xDS, falling back to other
+   * credentials if no configuration is provided.
+   *
+   * @throws IllegalArgumentException if fallback is unable to be used
+   */
+  public static ChannelCredentials createWithFallback(
+      ChannelCredentials fallback) {
+    return new InternalNettyChannelCredentials.create(
+        new XdsProtocolNegotiatorFactory(
+          InternalNettyChannelCredentials.toNegotiator(fallback)));
+  }
+}
+```
+
+### Netty ChannelCredentials Processing Example
+
+The API provided here is not part of the design. It is meant as a
+demonstration.
+
+```java
+final class ProtocolNegotiators {
+  private static final EnumSet<> understoodTlsFeatures =
```

Comment:
I'm having a hard time thinking of what a Feature would be.  Several of the TLS choices I have seen are more complex than an enum.

Reply:
To make it more concrete, one feature could be "MTLS" or "CLIENT_CERT". It would mean the caller knows how to call getCertificateChain(), getPrivateKey(), getPassword() to use the client cert. (Those methods would be added when we add the Feature.) Basically, as part of each Feature would be a specification of what it implies and the expected behavior, and there's no real limits to what could be specified.